x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/esnative/NativeUsersStore.java

```diff
@@ -309,6 +316,17 @@ public void updateReservedUser(
         });
     }
 
+    void storeAutoconfiguredElasticUser(ReservedUserInfo elasticUserInfo, ActionListener<ReservedUserInfo> listener) {
+        updateReservedUser(ElasticUser.NAME, elasticUserInfo.passwordHash, DocWriteRequest.OpType.CREATE,
+            WriteRequest.RefreshPolicy.IMMEDIATE, (e) -> {
+            if (e instanceof VersionConflictEngineException) {
+                listener.onFailure(e);
+            } else {
+                listener.onFailure(new ElasticsearchStatusException(e.getMessage(), INTERNAL_SERVER_ERROR, e.getCause()));
+            }
+            }, listener.delegateFailure((l, v) -> l.onResponse(elasticUserInfo.deepClone())));
+    }
```

Comment:
The logic here is wrong, I did some testing.
Firstly, the branching on the instance type should be on the `listener::onFailure` argument, not on the `consumer`. But even then, intercepting `VersionConflictEngineException` exceptions in order to return 401 instead of 500 is not correct. It is possible that multiple concurrent requests **on the same node** try to set the auto configured elastic user password symultaneously, and only one will succeed, but the others will be returned with 401 which is wrong (should signal authn success).

This is tricky I need to think some more for the proper fix.

Reply:
From https://github.com/elastic/elasticsearch/pull/77036#discussion_r708841870 , I think we should return 503 (slightly more suggestive than 500, I think) when the promised password cannot be persisted. This is what we set about to do, though back then I haven't considered that any other failures return 401.
Also, in particular, returning 401 because the elastic user becomes present in the .security index in the meantime, is technically wrong because it assumes the update must be with a different password, which might be true in practice, but only because of reserved realm caching, which means that the update comes from a different node, or the user, which probably use a different password to set for elastic, but not necessarily.